File: cpp/53_combinatoric_selections.cpp
#include <iostream>

int combinatoric_selections(){
    int count = 0;
    for (int n = 23; n <= 100; n++){
        int ncr = n;
        for (int r = 2; r <= n; r++){
            // nC1 = n
            // nC2 = n * (n - 1) / 2
            // nC3 = n * (n - 1) * (n - 2) / (2  * 3)
            // nCr = nC(r - 1) * (n - r + 1) / r
            ncr = ncr * (n - r + 1) / r;
            
            // nCr = nC(n - r) e.g. 23C10 = 23C13 > 1000000
            // r, (r + 1), (r + 2), ..., (n - r)
            // total of n - r - r + 1 = n - 2r + 1 values of r where nCr exceeds 100
            if (ncr > 1'000'000){
                count += n - 2*r + 1;
                break;
            }
        }
    }
    return count;
}

int main(){
    std::cout << combinatoric_selections() << "\n";
}

File: c#/_53_CombinatoricSelections.cs
using System;

public class _53_CombinatoricSelections{
	public static int Euler53(){
		int res = 0;
		for (int n = 23; n <= 100; n++){
			// nC1 = n! / ((n - 1)! * 1!) = n
			// nC2 = n! / ((n - 2)! * 2!) = n * (n - 1) / 2
			// nC3 = n! / ((n - 3)! * 3!) = n * (n - 1) * (n - 2) / 6
			// nCr = nC(r - 1) * (n - r + 1) / r
			int comb = 1;
			for (int r = 1; r <= n; r++){
				comb = comb * (n - r + 1) / r;
				if (comb > 1_000_000){
					// [r, n - r]
					// (n - r) - (r) + 1
					res += n - 2 * r + 1;
					break;	
				}
			}
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler53());
	}
}
